Notifications: More notification types, more settings per notification type

Theodlz opened the pull request:
### This PR modifies most of the current notification framework. Currently, it is not really possible to choose if you want to be notified by email, SMS, Slack, and **WHEN** to be notified **FOR EACH** notification type.  Also, a user can't be notified on some things that could be interesting (new gcn events **for specific notice types**, new sources, or new activity on sources **for certain classifications** ...).

#### This PR adds all of the above in the user preferences, as well as the required changes in the UserNotification model to get those new notifications to work.

#### In the user preferences, for each notification type (that are now: sources, gcn events, favorite sources, facility transactions, mentions), you have a Notification Setting Select menu, that allows you to choose to be notified: by email, by SMS (when you are on shift, and/or during a specific time period), by Slack.

### How to test the new features:

#### I would suggest reviewers to contact me on Slack when they want to review this PR, so I can send them my twillio and smtp config, as well as add them to the slack group I created to test the slack integration. That way it will be easier for them and less time-consuming.
#### Then, I suggest starting the app with an empty DB, selecting notifications on all fermi gcn notice types, activated notifications by email, SMS, and slack, with the right time slot (in UTC, I need to add that to the frontend), then loading the demo data and checking if they received notifications for the added gcn event. Then, they can add a source as a favorite, try the different notification types for favorite sources, and select a classification for the sources notification type to try that as well.

#### Also, maybe running the tests (that only test the in-app notifications btw) is the best way to understand what the new features are supposed to do.

### Status : 
#### I wrote frontend tests for every notification type where I verify a user received in-app notifications after enabling them. Also, I wrote a test to verify that the NotificationSettingsSelect component (where you select email, Slack, SMS, SMS on shift, SMS on time slot, adjust time slot...) works correctly, and saves the settings as expected (they are kept when refreshing the profile).

### What needs to be done:
#### Deciding the notification texts, for in-app, SMS, email, and Slack notifications, as well as the headers for emails. I would like this to be done with the reviewers, as they will use this feature, they are the best judges for that aspect of the PR. 
#### Some documentation ?

### PREVIEW
![notification_framework_beta](https://user-images.githubusercontent.com/49785117/174940017-306527e5-1620-46c4-bfeb-09da5e11eef8.gif)

Theodlz commented:
Hi @mcoughlin, tagging you first to have your opinion on the changes I made since you tested it.

mcoughlin commented on skyportal/models/user_notification.py:
```diff
@@ -21,9 +22,28 @@
 
+log = make_log('notifications')
+
+account_sid = cfg["twilio.sms_account_sid"]
```
Will this still work for whatever is happening here? https://github.com/skyportal/skyportal/pull/3091

Theodlz replied:
What do you mean by that ? I mean it should work, the key exists, but it has no value by default. What should we do, add a default null or ~ value in the config ? My PR aside, that's a problem in general with empty fields in the config.

mcoughlin replied:
Yeah I think it's fine.

mcoughlin commented on skyportal/models/user_notification.py:
```diff
@@ -21,9 +22,28 @@
+    client = TwilioClient(account_sid, auth_token)
+
+email = False
+if cfg["email_service"] == "sendgrid" or cfg["email_service"] == "smtp":
```
Should we have a separate flag in the config like enabling email? I.e. you can have it all setup fine but just turn it off with a email=false?

Theodlz replied:
Right now it's what the email service is for kinda, if you don't specify one (that's what is done by default) it won't use it, and otherwise, it will. But maybe an extra flag to just deactivate it without having to modify your email config could be a nice addition?

mcoughlin replied:
Yeah that's what I am thinking!

mcoughlin commented on skyportal/models/user_notification.py:
```diff
@@ -90,30 +114,172 @@ def send_slack_notification(mapper, connection, target):
     slack_microservice_url = (
         f'http://127.0.0.1:{cfg.get("slack.microservice_port", 64100)}'
     )
+    ressource_type = None
```
resource

mcoughlin commented on skyportal/models/user_notification.py:
```diff
@@ -90,30 +114,172 @@ def send_slack_notification(mapper, connection, target):
+@event.listens_for(UserNotification, 'after_insert')
+def send_email_notification(mapper, connection, target):
+
+    if not email:
```
should we combine some of these as like a long set in one if?

mcoughlin commented on skyportal/models/user_notification.py:
```diff
@@ -90,30 +114,172 @@ def send_slack_notification(mapper, connection, target):
+
+    ressource_type = None
+    if "favorite_sources" not in target.notification_type:
+        ressource_type = target.notification_type
```
resource :)

mcoughlin commented on skyportal/models/user_notification.py:
```diff
@@ -90,30 +114,172 @@ def send_slack_notification(mapper, connection, target):
+@event.listens_for(UserNotification, 'after_insert')
+def send_sms_notification(mapper, connection, target):
+    if client is None:
+        return
```
Can these if blocks be reused between notification types?

Theodlz replied:
You're right, some of them can be reused. I'll add a separate function then.

Theodlz replied:
Hmmm, I'm actually not so sure if it's worth doing. Most of those if are different actually. Not os many of them can be reused. Depending on the notification type we check different stuff.

Theodlz replied:
checking if the user has preferences and such is redundant, so I'll create a function at least for that.

Theodlz replied:
Ok that's done.

mcoughlin reviewed (commented):
@Theodlz Really nice work so far, and the interface is really good. My only substantive comment is that if there are ways to simplify the "to notify or not to notify, that is the question" type logic, including making a function to reuse between the different notification types, that might be nice. On the other hand, all this stuff is almost necessarily convoluted, so I thought you did a nice job.

guynir42 commented:
I'm going to review now. Thanks for clearing all of this up. And good work on the PR description, this makes it a lot easier for me to know what I'm going to be looking at. 
Let me know if you also want me to test locally with the email and SMS alerts.

guynir42 commented on skyportal/handlers/api/gcn.py:
```diff
@@ -577,3 +579,17 @@ def delete(self, dateobs, localization_name):
         return self.success()
+
+
+class GcnNoticeTypesHandler(BaseHandler):
+    @auth_or_token
+    def get(self):
+        '''
+        This method return the list of notices from the config file
+        ---
+        description: Returns the list of notices
+        tags:
+          - gcn_notices
+        '''
+        notices = cfg['gcn_notice_types']
+        return self.success(data=notices)
```
I think you may be able to get this without using a separate handler. Look at `static/js/ducks/config.js` and `skyportal/handlers/api/config_handler.py` .

Theodlz replied:
Oh that's great ! I've completely missed that config_handler. I'll try using that instead.

Theodlz replied:
Done.

guynir42 commented on static/js/components/GcnNoticeTypesSelect.jsx:
```diff
@@ -0,0 +1,43 @@
+const GcnNoticeTypesSelect = (props) => {
+  const { selectedGcnNoticeTypes, setSelectedGcnNoticeTypes } = props;
+  const dispatch = useDispatch();
+  const gcn_notice_types = useSelector(
+    (state) => state.gcnNoticeTypes.gcnNoticeTypes
+  );
+
+  useEffect(() => {
+    dispatch(fetchGcnNoticeTypes());
+  }, []);
```
If you add this to the `config_handler` it should be loaded into the state when the app is "hydrated".

Theodlz replied:
Sounds good !

guynir42 reviewed (commented):
Code looks good. Let me know if you would like to close the loop on testing the actual notifications.

Theodlz commented:
> I'm going to review now. Thanks for clearing all of this up. And good work on the PR description, this makes it a lot easier for me to know what I'm going to be looking at. Let me know if you also want me to test locally with the email and SMS alerts.

Hi @guynir42 . Thanks for the review ! I'll send you the twillio config, slack URL and add you to the slack group i created so you can test all of this without having to do everything from scratch.

guynir42 commented:
I'm getting some console errors and warnings when playing with the user preferences: 
```
 A component is changing the uncontrolled checked state of SwitchBase to be controlled.
Elements should not switch from uncontrolled to controlled (or vice versa).
Decide between using a controlled or uncontrolled SwitchBase element for the lifetime of the component.
The nature of the state is determined during the first render. It's considered controlled if the value is not `undefined`.
```

guynir42 commented:
And also this warning: 
```
You have provided an out-of-range value `undefined` for the select (name="followupRequestAllocationSelect") component.
Consider providing a value that matches one of the available options or ''.
The available values are `-1`, `1`, `2`, `3`, `4`, `5`, `6`, `7`, `8`, `9`, `10`, `11`, `12`, `13`, `14`, `15`, `16`.
```

guynir42 commented:
First impressions from playing with the new notifications: I think this is very cool and so far I've been able to get slack messages. 

I would add tooltips on **everything** in the user preferences page. It is not immediately obvious what the buttons do (especially the notification icon).

Theodlz commented:
Hi @guynir42, I added the tooltips you asked for.

> First impressions from playing with the new notifications: I think this is very cool and so far I've been able to get slack messages.
> 
> I would add tooltips on **everything** in the user preferences page. It is not immediately obvious what the buttons do (especially the notification icon).

Hi @guynir42, I added the tooltips you asked for.

mcoughlin reviewed (approved):
@Theodlz Thanks for fixing all that. This is going to be really good.
